File: common/src/main/org/avis/common/ElvinURI.java
package org.avis.common;

import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import java.util.Map.Entry;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import java.net.URISyntaxException;


import static java.lang.Integer.parseInt;
import static java.util.Arrays.asList;
import static java.util.Collections.emptyMap;

import static org.avis.common.Common.CLIENT_VERSION_MAJOR;
import static org.avis.common.Common.CLIENT_VERSION_MINOR;
import static org.avis.common.Common.DEFAULT_PORT;
import static org.avis.util.Collections.join;
import static org.avis.util.Collections.list;

/**
 * An Elvin URI identifying an Elvin router as described in the "Elvin
 * URI Scheme" specification at
 * http://elvin.org/specs/draft-elvin-uri-prelim-02.txt. The most
 * common Elvin URI for a TCP endpoint is of the form:
 * 
 * <pre>
 *   elvin:[version]/[protocol]/hostname[:port];[options]
 *   
 *   protocol: transport,security,marshalling
 *   options:  name1=value1[;name2=value2]*
 * </pre>
 * 
 * <p>
 * 
 * Example URI 1: <code>elvin://localhost:2917</code>.
 * <p>
 * Example URI 2: <code>elvin:4.0/tcp,ssl,xdr/localhost:29170</code>.
 * 
 * todo need to support []'s in IPv6 addresses
 * 
 * @author Matthew Phillips
 */
public final class ElvinURI
{
  private static final List<String> DEFAULT_PROTOCOL =
    list ("tcp", "none", "xdr");
  
  /*
   * Note: you might expect "tcp,ssl,xdr" as the logical secure stack,
   * but Mantara Elvin uses "ssl,none,xdr" and spec also indicates
   * this is correct, so we comply.
   */
  private static final List<String> SECURE_PROTOCOL =
    list ("ssl", "none", "xdr");
  
  /**
   * Basic matcher for URI's. Detail parsing is done as a separate pass.
   */
  private static final Pattern URL_PATTERN =
    Pattern.compile ("(\\w+):(?:([^/]+))?/([^/]+)?/([^/].*?)(;.*)?");

  /**
   * The URI string as passed into the constructor.
   */
  public String uriString;
  
  /**
   * Major protocol version. Default {@link Common#CLIENT_VERSION_MAJOR}.
   */
  public int versionMajor;
  
  /**
   * Minor protocol version. Default {@link Common#CLIENT_VERSION_MINOR}.
   */
  public int versionMinor;
  
  /**
   * The stack of protocol modules in (transport,security,marshalling)
   * order. e.g. "tcp", "none", "xdr". See also
   * {@link #defaultProtocol()}
   */
  public List<String> protocol;
  
  /**
   * Host name.
   */
  public String host;
  
  /**
   * Port. Default {@link Common#DEFAULT_PORT}.
   */
  public int port;

  /**
   * URI options: e.g. elvin://host:port;option1=value1;option2=value2
   */
  public Map<String, String> options;

  private int hash;

  /**
   * Create a new instance.
   * 
   * @param uriString The URI.
   * 
   * @throws URISyntaxException if the URI is not valid.
   */
  public ElvinURI (String uriString)
    throws URISyntaxException
  {
    init ();

    this.uriString = uriString;
    
    parseUri ();
    
    this.hash = computeHash ();
  }

  /**
   * Create a new instance from a host and port using defaults for others.
   * 
   * @param host Host name or IP address
   * @param port Port number.
   */
  public ElvinURI (String host, int port)
  {
    init ();
    
    this.uriString = "elvin://" + host + ':' + port;
    this.host = host;
    this.port = port;
    this.hash = computeHash ();
  }

  /**
   * Create a new instance using an existing URI for defaults.
   * 
   * @param uriString The URI string.
   * @param defaultUri The URI to use for any values that are not
   *          specified by uriString.
   * @throws URISyntaxException if the URI is not valid.
   */
  public ElvinURI (String uriString, ElvinURI defaultUri)
    throws URISyntaxException
  {
    init (defaultUri);
    
    this.uriString = uriString;
    
    parseUri ();
    
    this.hash = computeHash ();
  }
  
  /**
   * Create a copy of a URI.
   * 
   * @param defaultUri The URI to copy.
   */
  public ElvinURI (ElvinURI defaultUri)
  {
    init (defaultUri);
  }

  private void init (ElvinURI defaultUri)
  {
    this.uriString = defaultUri.uriString;
    this.versionMajor = defaultUri.versionMajor;
    this.versionMinor = defaultUri.versionMinor;
    this.protocol = defaultUri.protocol;
    this.host = defaultUri.host;
    this.port = defaultUri.port;
    this.options = defaultUri.options;
    this.hash = defaultUri.hash;
  }

  private void init ()
  {
    this.versionMajor = CLIENT_VERSION_MAJOR;
    this.versionMinor = CLIENT_VERSION_MINOR;
    this.protocol = DEFAULT_PROTOCOL;
    this.host = null;
    this.port = DEFAULT_PORT;
    this.options = emptyMap ();
  }

  @Override
  public String toString ()
  {
    return uriString;
  }
  
  public String toCanonicalString ()
  {
    StringBuilder str = new StringBuilder ();
    
    str.append ("elvin:");
    
    str.append (versionMajor).append ('.').append (versionMinor);
    
    str.append ('/');
    
    join (str, protocol, ',');
    
    str.append ('/').append (host).append (':').append (port);

    // NB: options is a sorted map, canonical order is automatic
    for (Entry<String, String> option : options.entrySet ())
    {
      str.append (';');
      str.append (option.getKey ()).append ('=').append (option.getValue ());
    }
    
    return str.toString ();
  }
  
  @Override
  public int hashCode ()
  {
    return hash;
  }
  
  @Override
  public boolean equals (Object obj)
  {
    return obj instanceof ElvinURI && equals ((ElvinURI)obj);
  }

  public boolean equals (ElvinURI uri)
  {
    return hash == uri.hash &&
           host.equals (uri.host) &&
           port == uri.port &&
           versionMajor == uri.versionMajor &&
           versionMinor == uri.versionMinor && 
           options.equals (uri.options) &&
           protocol.equals (uri.protocol);
  }
  
  private int computeHash ()
  {
    return host.hashCode () ^ port ^ protocol.hashCode ();
  }
  
  private void parseUri ()
    throws URISyntaxException
  { 
    Matcher matcher = URL_PATTERN.matcher (uriString);
    
    if (!matcher.matches ())
    {
      throw new URISyntaxException (uriString, "Not a valid Elvin URI");
    } else if (!matcher.group (1).equals ("elvin"))
    {
      throw new URISyntaxException (uriString,
                                    "Elvin URI scheme must be \"elvin:\"");
    }
    
    // version
    if (matcher.group (2) != null)
      parseVersion (matcher.group (2));
    
    // protocol
    if (matcher.group (3) != null)
      parseProtocol (matcher.group (3));
    
    // endpoint (host/port)
    parseEndpoint (matcher.group (4));
    
    // options
    if (matcher.group (5) != null)
      parseOptions (matcher.group (5));
  }

  private void parseVersion (String versionExpr)
    throws URISyntaxException
  {
    Matcher versionMatch =
      Pattern.compile ("(\\d+)(?:\\.(\\d+))?").matcher (versionExpr);
    
    if (versionMatch.matches ())
    {
      try
      {
        versionMajor = parseInt (versionMatch.group (1));
        
        if (versionMatch.group (2) != null)
          versionMinor = parseInt (versionMatch.group (2));
      } catch (NumberFormatException ex)
      {
        throw new URISyntaxException (uriString,
                                      "Number too large in version string: \"" +
                                      versionExpr + "\"");
      }
    } else
    {
      throw new URISyntaxException (uriString,
                                    "Invalid version string: \"" +
                                    versionExpr + "\"");
    }
  }
  
  private void parseProtocol (String protocolExpr)
    throws URISyntaxException
  {
    Matcher protocolMatch =
      Pattern.compile ("(?:(\\w+),(\\w+),(\\w+))|secure").matcher (protocolExpr);
    
    if (protocolMatch.matches ())
    {
      if (protocolMatch.group (1) != null)
        protocol = asList (protocolExpr.split (","));
      else
        protocol = SECURE_PROTOCOL;
    } else
    {
      throw new URISyntaxException (uriString,
                                    "Invalid protocol: \"" +
                                    protocolExpr + "\"");
    }
  }
  
  private void parseEndpoint (String endpoint)
    throws URISyntaxException
  {
    Matcher endpointMatch =
      Pattern.compile ("([^:]+)(?::(\\d+))?").matcher (endpoint);
    
    if (endpointMatch.matches ())
    {
      host = endpointMatch.group (1);
      
      if (endpointMatch.group (2) != null)
        port = parseInt (endpointMatch.group (2));
    } else
    {
      throw new URISyntaxException (uriString, "Invalid port number");
    }
  }
  
  private void parseOptions (String optionsExpr)
    throws URISyntaxException
  {
    Matcher optionMatch =
      Pattern.compile (";([^=;]+)=([^=;]*)").matcher (optionsExpr);
    
    options = new TreeMap<String, String> ();
    
    int index = 0;
    
    while (optionMatch.lookingAt ())
    {
      options.put (optionMatch.group (1), optionMatch.group (2));
      
      index = optionMatch.end ();
      optionMatch.region (index, optionsExpr.length ());
    }
    
    if (index != optionsExpr.length ())
      throw new URISyntaxException
        (uriString, "Invalid options: \"" + optionsExpr + "\"");
  }

  /**
   * The default URI protocol stack: "tcp", "none", "xdr"
   */
  public static List<String> defaultProtocol ()
  {
    return DEFAULT_PROTOCOL;
  }

  /**
   * The secure URI protocol stack: "ssl", "none", "xdr"
   */
  public static List<String> secureProtocol ()
  {
    return SECURE_PROTOCOL;
  }
}

File: branches/sandbox/mina_2.0M1_conversion/cpp/common/src/main/ElvinURI.cpp
#include "ElvinURI.h"
#include <sstream>
#include "pcrecpp.h"

ElvinURI::ElvinURI(char *theuriString)
{
	DEFAULT_PROTOCOL.push_back("tcp");
	DEFAULT_PROTOCOL.push_back("none");
	DEFAULT_PROTOCOL.push_back("xdr");
	
	SECURE_PROTOCOL.push_back("ssl");
	SECURE_PROTOCOL.push_back("none");
	SECURE_PROTOCOL.push_back("xdr");
	
	init();
	uriString = theuriString;
	parseUri();
	validate();
}

ElvinURI::ElvinURI(char *host, int port)
: m_port(port)
{
	DEFAULT_PROTOCOL.push_back("tcp");
	DEFAULT_PROTOCOL.push_back("none");
	DEFAULT_PROTOCOL.push_back("xdr");
	
	SECURE_PROTOCOL.push_back("ssl");
	SECURE_PROTOCOL.push_back("none");
	SECURE_PROTOCOL.push_back("xdr");

	init();
	uriString.append("elvin://");
	uriString.append(host);
	uriString.append(":");

	std::ostringstream oss;
	oss << port;
	uriString.append(oss.str());

	scheme = "elvin";
	m_host = host;
}

ElvinURI::ElvinURI(char *theuriString, ElvinURI &defaultUri)
{
	DEFAULT_PROTOCOL.push_back("tcp");
	DEFAULT_PROTOCOL.push_back("none");
	DEFAULT_PROTOCOL.push_back("xdr");
	
	SECURE_PROTOCOL.push_back("ssl");
	SECURE_PROTOCOL.push_back("none");
	SECURE_PROTOCOL.push_back("xdr");

	init(defaultUri);
	uriString = theuriString;
	parseUri();
	validate();
}

ElvinURI::ElvinURI(ElvinURI &defaultUri)
{
	DEFAULT_PROTOCOL.push_back("tcp");
	DEFAULT_PROTOCOL.push_back("none");
	DEFAULT_PROTOCOL.push_back("xdr");
	
	SECURE_PROTOCOL.push_back("ssl");
	SECURE_PROTOCOL.push_back("none");
	SECURE_PROTOCOL.push_back("xdr");

	init(defaultUri);
	validate();
}

ElvinURI::~ElvinURI()
{
}

void ElvinURI::init(ElvinURI &defaultUri)
{
	uriString = defaultUri.uriString;
	scheme = defaultUri.scheme;
	versionMajor = defaultUri.versionMajor;
	versionMinor = defaultUri.versionMinor;
	protocol = defaultUri.protocol;
	m_host = defaultUri.m_host;
	m_port = defaultUri.m_port;
	options = defaultUri.options;
}

void ElvinURI::init()
{
	scheme = "";
	versionMajor = CLIENT_VERSION_MAJOR;
	versionMinor = CLIENT_VERSION_MINOR;
	protocol = DEFAULT_PROTOCOL;
	m_host = "";
	m_port = DEFAULT_PORT;
}

void ElvinURI::validate()
{
	if (!validScheme(scheme.c_str()))
	{
		std::string errormessage;
		errormessage.append("Invalid scheme: ");
		errormessage.append(scheme.c_str());
		throw InvalidURIException(uriString.c_str(), errormessage.c_str());
	}
}

bool ElvinURI::validScheme(const char *schemeToCheck)
{	
	if (strcmp(schemeToCheck, "elvin") == 0)
		return true;
	else
		return false;
}

std::string ElvinURI::toString()
{
	return uriString;
}

/**
* Generate a canonical text version of this URI.
*/
std::string ElvinURI::toCanonicalString()
{
	std::string str;
	str.append(scheme);
	str.append(":");

	std::ostringstream vmajoss;
	vmajoss << versionMajor;
	str.append(vmajoss.str());

	str.append(".");
	
	std::ostringstream vminoss;
	vminoss << versionMinor;
	str.append(vminoss.str());

	str.append("/");

	bool first = true;
	std::list<std::string>::iterator listiter;
	listiter = protocol.begin();
	while (listiter != protocol.end())
	{	
		if (!first)
		{
			str.append(",");
		}
		first = false;
		str.append(*listiter);
		listiter++;
	}
	
	str.append("/");
	str.append(m_host);
	str.append (":");

	std::ostringstream oss;
	oss << m_port;

	str.append(oss.str());

	// NB: options is a sorted map, canonical order is automatic
	std::map<std::string, std::string>::iterator iter;
	iter = options.begin();
	while (iter != options.end())
	{
		str.append(";");
		str.append(iter->first);
		str.append("=");
		str.append(iter->second);
		iter++;
	}

	return str;
}

bool ElvinURI::equals(ElvinURI &uri)
{
	return ((scheme.compare(uri.scheme) == 0 ? true : false) &&
		(m_host.compare(uri.m_host) == 0 ? true : false) &&
		(m_port == uri.m_port) &&
		(versionMajor == uri.versionMajor) &&
		(versionMinor == uri.versionMinor) &&
		(options == uri.options) &&
		(protocol == uri.protocol));
}

void ElvinURI::parseUri()
{ 
	string group1, group2, group3, group4, group5;
	pcrecpp::RE pat("(\\w+):([^/]*)/([^/]*)/([^;/][^;]*)(;.*)?");
	if (pat.FullMatch(uriString, &group1, &group2, &group3, &group4, &group5))
	{
		scheme = group1;

		// version
		if (group2.length() != 0)
		{
			parseVersion(group2.c_str());
		}

		// protocol
		if (group3.length() != 0)
		{
			parseProtocol(group3.c_str());
		}

		// endpoint (host/port)
		parseEndpoint(group4.c_str());

		// options
		if (group5.length() != 0)
		{
			parseOptions(group5.c_str());
		}
	}
	else
	{
		throw InvalidURIException(uriString.c_str(), "Not a valid Elvin URI");
	}
}

void ElvinURI::parseVersion(const char *versionExpr)
{
	string group1, group2, group3;
	pcrecpp::RE pat("(\\d+)(\\.(\\d+))?");
	if (pat.FullMatch(versionExpr, &group1, &group2, &group3))
	{
		std::istringstream versionMajorStream(group1.c_str());
		if (!(versionMajorStream >> versionMajor))
		{
			std::string errormessage;
			errormessage.append("Invalid version string: \"");
			errormessage.append(versionExpr);
			errormessage.append("\"");
			throw InvalidURIException(uriString.c_str(), errormessage.c_str());
		}

		if (group3.length() > 0)
		{
			std::istringstream versionMinorStream(group3.c_str());
			if (!(versionMinorStream >> versionMinor))
			{
				std::string errormessage;
				errormessage.append("Invalid version string: \"");
				errormessage.append(versionExpr);
				errormessage.append("\"");
				throw InvalidURIException(uriString.c_str(), errormessage.c_str());			
			}
		}
	} 
	else
	{
		std::string errormessage;
		errormessage.append("Invalid version string: \"");
		errormessage.append(versionExpr);
		errormessage.append("\"");
		throw InvalidURIException(uriString.c_str(), errormessage.c_str());
	}
}

void ElvinURI::parseProtocol(const char *protocolExpr)
{
	string group1, group2, group3, group4;
	pcrecpp::RE pat("((\\w+),(\\w+),(\\w+))|secure");
	if (pat.FullMatch(protocolExpr, &group1, &group2, &group3, &group4))
	{
		if (group2.length() > 0)
		{
			//clear current protocol list first before adding new entries
			protocol.clear();

			protocol.push_back(group2.c_str());
			protocol.push_back(group3.c_str());
			protocol.push_back(group4.c_str());			
		}
		else
		{	
			protocol = SECURE_PROTOCOL;
		}
	} 
	else
	{
		std::string errormessage;
		errormessage.append("Invalid protocol: \"");
		errormessage.append(protocolExpr);
		errormessage.append("\"");
		throw InvalidURIException(uriString.c_str(), errormessage.c_str());
	}
}

void ElvinURI::parseEndpoint(const char *endpoint)
{
	std::string patternstring;

	// choose between IPv6 and IPv4 address scheme
	if (endpoint[0] == '[')
		patternstring = "(\\[[^\\]]+\\])(:(\\d+))?";
	else
		patternstring =  "([^:]+)(:(\\d+))?";

	string group1, group2, group3;
	pcrecpp::RE pat(patternstring);
	if (pat.FullMatch(endpoint, &group1, &group2, &group3))
	{
		m_host = group1;

		if (group3.length() > 0)
			m_port = atoi (group3.c_str());
	} 
	else
	{
		throw InvalidURIException(uriString.c_str(), "Invalid port number");
	}
}

void ElvinURI::parseOptions(const char *optionsExpr)
{
	pcrecpp::StringPiece input(optionsExpr);

	string group1, group2;
	pcrecpp::RE pat(";([^=;]+)=([^=;]*)");
	int matchcount = 0;
	while (pat.Consume(&input, &group1, &group2)) 
	{
		if (group1.length() > 0)
		{
			if (group1.length() > 0)
			{
				matchcount++;
				options.insert(std::pair<std::string, std::string> (group1, group2));
			}
			else
			{
				std::string errormessage;
				errormessage.append("Invalid options: \"");
				errormessage.append(optionsExpr);
				errormessage.append("\"");
				throw InvalidURIException(uriString.c_str(), errormessage.c_str());
			}
		}
	}
	
	if (matchcount == 0)
	{
		std::string errormessage;
		errormessage.append("Invalid options: \"");
		errormessage.append(optionsExpr);
		errormessage.append("\"");
		throw InvalidURIException(uriString.c_str(), errormessage.c_str());
	}
}

/**
* The default URI protocol stack: "tcp", "none", "xdr"
*/
std::list<std::string> ElvinURI::defaultProtocol()
{
	return DEFAULT_PROTOCOL;
}

/**
* The secure URI protocol stack: "ssl", "none", "xdr"
*/
std::list<std::string> ElvinURI::secureProtocol()
{
	return SECURE_PROTOCOL;
}
